### PR TITLE
add kms decryption support for secrets via sops

### PR DESCRIPTION
**Pull request by emalloy**
given an envar `SOPS_KMS_ARN` and items in the `SOPS_SECRETS` array; k8s-scripts will decrypt a named item `<item>.secret.sops.yml` prior to executing `kubectl apply -f` during a deployment of a secret.

the tool, sops, has code and documentation hosted here: https://github.com/mozilla/sops

and this practice is largely based on a [blog post](https://frederic-hemberger.de/articles/manage-kubernetes-secrets-with-sops/)

with this, it could be said that committing an encrypted file `<item>.secret.sops.yml`  is safe.

the process for encrypting the secrets spec file is such:

1. create a [kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/) spec file
1. `export SOPS_KMS_ARN` - the value being the ARN of the KMS key to encrypt with. *or* just specify the arn using the `--kms` flag on invocation
1. `sops --kms="<your kms key>" --encrypt mysecret.yaml > mysecret.secret.sops.yaml` 
1. Commit and define in your `<namespace>.config`

### CODE REVIEW

File: bin/k8s-read-config [hunk 1]
```diff
@@ -59,6 +59,15 @@ echo "[secret_files]: ${SECRET_FILES[@]}"
...
 echo "[s3-secrets]: ${S3_SECRETS[@]}"
 
+
+echo "[sops-secrets]: ${SOPS_SECRETS[@]}"
+SOPS_SECRET_FILES=()
+for i in "${SOPS_SECRETS[@]}"
+do
+  SOPS_SECRET_FILES+=($BASEDIR/deploy/$i.secret.sops.yml)
+done
+echo "[secret_files]: ${SOPS_SECRET_FILES[@]}"
```
Comment (philipcristiano) [inline bin/k8s-read-config]: Should `secret_files` be `sops_secret_files` to avoid similar lines?

File: README.md [hunk 2]
```diff
@@ -192,10 +194,44 @@ of all services that are accessible from your local machine
 Makes sure kubectl is installed and available for use. Customize the version
 by specifying the `KUBECTL_VERSION` envrionmental variable. Default: `v1.3.6`.
 
+
 ## Assumptions
 
 * In your Deployment file, specify imagePullPolicy: IfNotPresent
 
+# Errata
```
Comment (philipcristiano) [inline README.md]: `Errata` may not be the correct word here, this doesn't appear to a correction. 

This section seems like it should be included as docs for the deploy script based on the current implementation.

**Comment by philipcristiano** [comment]
This is really cool! 

This behaves differently than the s3-secrets by including these in the deploy script instead of a separate script. We should try to keep the semantics of these scripts as similar as possible. 

I'm a bit concerned about the failure mode of a deploy if sops errors occur. The deploy will partially succeed for configmaps and secrets, but not sops-secrets. There isn't a separate way to test this before making changes in the Kube cluster. This could cause problems depending on the length of time required to fix the deployment errors.

I like that it avoids writing secrets to disk!  

If we go with this path it would be good to make a breaking change and move secrets generation to the top of the deploy script and store in memory. That way a failure on 3rd party components wouldn't stop the deployment part of the way through and we'd avoid writing secrets to disk (other than in Kube). What do you think?

**Comment by emalloy** [comment]
@philipcristiano Thanks for taking a look - and excellent catch on the error handling. Also agree with your suggestion regarding keeping the deployment commands as similar as possible -- so I will see how I can break out the decryption + pipe to kubectl into a separate script. I might be a little behind on this today, but will get to it first thing monday.

**Comment by emalloy** [comment]
This was refactored based on @philipcristiano 's recommendations, however, following a conversation between Philip and I, it's going to be further refactored for better uniformity with how other secret decryption mechanisms are used. Additionally, I had just run further tests and secrets are not being decrypted as expected. 

Do not merge.

**Comment by emalloy** [comment]
This PR is ready for final review. Sorry for the delay

**Comment by philipcristiano** [comment]
LGTM, great work!

**Pull request merged**